fem/kernels.hpp

```diff
@@ -60,183 +148,882 @@ MFEM_HOST_DEVICE inline void LoadX(const int e, const int D1D, const int c,
+   MFEM_SYNC_THREAD;
+}
+
+/// Pull 2D Gradient
+template<int MQ1, int NBZ>
+MFEM_HOST_DEVICE inline void PullGrad(const int qx, const int qy,
+                                      const double (&sQQ)[4][NBZ][MQ1*MQ1],
+                                      double *Jpr)
+{
+   const int tidz = MFEM_THREAD_ID(z);
+   ConstDeviceMatrix X0GB(sQQ[0][tidz], MQ1, MQ1);
+   ConstDeviceMatrix X0BG(sQQ[1][tidz], MQ1, MQ1);
+   ConstDeviceMatrix X1GB(sQQ[2][tidz], MQ1, MQ1);
+   ConstDeviceMatrix X1BG(sQQ[3][tidz], MQ1, MQ1);
+
+   Jpr[0] = X0GB(qx,qy);
+   Jpr[1] = X1GB(qx,qy);
+   Jpr[2] = X0BG(qx,qy);
+   Jpr[3] = X1BG(qx,qy);
+}
+
+/// Push 2D Gradient
+template<int MQ1, int NBZ>
+MFEM_HOST_DEVICE inline void PushGrad(const int qx, const int qy,
+                                      const double *A,
+                                      double (&sQQ)[4][NBZ][MQ1*MQ1])
+{
+   const int tidz = MFEM_THREAD_ID(z);
+   DeviceMatrix X0GB(sQQ[0][tidz], MQ1, MQ1);
+   DeviceMatrix X0BG(sQQ[1][tidz], MQ1, MQ1);
+   DeviceMatrix X1GB(sQQ[2][tidz], MQ1, MQ1);
+   DeviceMatrix X1BG(sQQ[3][tidz], MQ1, MQ1);
+
+   X0GB(qx,qy) = A[0];
+   X1GB(qx,qy) = A[2];
+   X0BG(qx,qy) = A[1];
+   X1BG(qx,qy) = A[3];
+}
+
+/// 2D Transposed gradient, 1/2
+template<int MD1, int MQ1, int NBZ>
+MFEM_HOST_DEVICE inline void GradYt(const int D1D, const int Q1D,
+                                    const double (&sBG)[2][MQ1*MD1],
+                                    const double (&GQ)[4][NBZ][MQ1*MQ1],
+                                    double (&GD)[4][NBZ][MD1*MQ1])
+{
+   const int tidz = MFEM_THREAD_ID(z);
+   ConstDeviceMatrix Bt(sBG[0], MQ1, MD1);
+   ConstDeviceMatrix Gt(sBG[1], MQ1, MD1);
+   ConstDeviceMatrix QQx0(GQ[0][tidz], MQ1, MQ1);
+   ConstDeviceMatrix QQx1(GQ[1][tidz], MQ1, MQ1);
+   ConstDeviceMatrix QQy0(GQ[2][tidz], MQ1, MQ1);
+   ConstDeviceMatrix QQy1(GQ[3][tidz], MQ1, MQ1);
+   DeviceMatrix DQxB(GD[0][tidz], MQ1, MD1);
+   DeviceMatrix DQxG(GD[1][tidz], MQ1, MD1);
+   DeviceMatrix DQyB(GD[2][tidz], MQ1, MD1);
+   DeviceMatrix DQyG(GD[3][tidz], MQ1, MD1);
+
+   MFEM_FOREACH_THREAD(qy,y,Q1D)
+   {
+      MFEM_FOREACH_THREAD(dx,x,D1D)
+      {
+         double u[2] = {0.0, 0.0};
+         double v[2] = {0.0, 0.0};
+         for (int qx = 0; qx < Q1D; ++qx)
+         {
+            u[0] += Gt(qx,dx) * QQx0(qx,qy);
+            u[1] += Gt(qx,dx) * QQy0(qx,qy);
+            v[0] += Bt(qx,dx) * QQx1(qx,qy);
+            v[1] += Bt(qx,dx) * QQy1(qx,qy);
+         }
+         DQxB(qy,dx) = u[0];
+         DQyB(qy,dx) = u[1];
+         DQxG(qy,dx) = v[0];
+         DQyG(qy,dx) = v[1];
+      }
+   }
+   MFEM_SYNC_THREAD;
+}
+
+/// 2D Transposed gradient, 2/2
+template<int MD1, int MQ1, int NBZ>
+MFEM_HOST_DEVICE inline void GradXt(const int D1D, const int Q1D,
+                                    const double (&sBG)[2][MQ1*MD1],
+                                    const double (&GD)[4][NBZ][MD1*MQ1],
+                                    const DeviceTensor<4> &Y,
```

Comment:
@v-dobrev You can witness how unintuitive the signature of this  function is because of non-`mutable` lambdas.
The captured `Y` is `const` due to non-`mutable` lambdas, however `Y` is the variable that this function modifies (thanks to the non-`const` `const` accessor of `DeviceTensor`), therefore this function **output** has to be `const` (?!).
I think `const` is very misleading here as it is mandatory **and** applied to the variable this function modifies.

Reply:
It may be confusing at a first glance, I agree. But then you remember that the `const` applies to the "pointer", not to the data, and it becomes clear. We can add a simple comment at the end of the line as a reminder: `// output param`.

Reply:
I've added `// output` comments where this kind of signature was needed.